ext/ffi/ffi.c

```diff
@@ -3907,6 +3907,9 @@ ZEND_METHOD(FFI, cast) /* {{{ */
 	 && type->kind == ZEND_FFI_TYPE_POINTER) {
 		cdata->ptr = &cdata->ptr_holder;
 		cdata->ptr_holder = old_cdata->ptr;
+	} else if (old_type->kind== ZEND_FFI_TYPE_POINTER
+	 && type->kind == ZEND_FFI_TYPE_ARRAY) {
+		cdata->ptr = old_cdata->ptr_holder;
```

Comment:
I think this should be extended with check for pointer/array type compatibility.
```diff
diff --git a/ext/ffi/ffi.c b/ext/ffi/ffi.c
index 352a732a49..e2a8a7f466 100644
--- a/ext/ffi/ffi.c
+++ b/ext/ffi/ffi.c
@@ -3901,9 +3901,14 @@ ZEND_METHOD(FFI, cast) /* {{{ */
 		/* automatically dereference void* pointers ??? */
 		cdata->ptr = *(void**)ptr;
 	} else if (old_type->kind == ZEND_FFI_TYPE_ARRAY
-	 && type->kind == ZEND_FFI_TYPE_POINTER) {
+	 && type->kind == ZEND_FFI_TYPE_POINTER
+	 && zend_ffi_is_compatible_type(ZEND_FFI_TYPE(old_type->array.type), ZEND_FFI_TYPE(type->pointer.type))) {
 		cdata->ptr = &cdata->ptr_holder;
 		cdata->ptr_holder = old_cdata->ptr;
+	} else if (old_type->kind == ZEND_FFI_TYPE_POINTER
+	 && type->kind == ZEND_FFI_TYPE_ARRAY
+	 && zend_ffi_is_compatible_type(ZEND_FFI_TYPE(old_type->pointer.type), ZEND_FFI_TYPE(type->array.type))) {
+		cdata->ptr = old_cdata->ptr_holder;
 	} else if (type->size > old_type->size) {
 		zend_object_release(&cdata->std);
 		zend_throw_error(zend_ffi_exception_ce, "attempt to cast to larger type");
```

I'm also not completely sure about ownership of the resulting array. Note, that the resulting array is actually points to elements of the original array. The following code will change $value.

```php
$a = FFI::cast('char[4]', $value + 4);
$a[0] = "X";
var_dump($a, $value);
```   

Probably, this is not a big problem if ``$value`` is not deallocated before ``$a``.

**C** doesn't allow to cast to arrays at all ``(char[4])ptr`` would lead to compile error. It allows casting to pointer to array ``(char(*)[4])ptr``.

Reply:
> I think this should be extended with check for pointer/array type compatibility.

Ah, right. Thanks!

> Probably, this is not a big problem if $value is not deallocated before $a.

I agree, and the [documentation](https://www.php.net/manual/en/ffi.cast.php) already states:

> FFI::cast() creates a new FFI\CData object, that references the same C data structure, but is associated with a different type. The resulting object does not own the C data, and the source ptr must survive the result.

> C doesn't allow to cast to arrays at all

Right. I don't think casting pointer to array isn't particularly useful regarding interfacing with C libraries, but rather for some other trickery; see also https://github.com/php/php-src/issues/7866#issuecomment-1005922117.

If we don't want to support casting pointer to array, we need to explicitly forbid it; as it's now, it's not properly implemented.